// plex/lib/tree.js
// Generated by CoffeeScript 1.4.0
var Tree;

Tree = (function() {

  function Tree(context) {
    this.context = context;
    this.edges = {
      local: {},
      remote: {}
    };
  }

  Tree.prototype.insertLocal = function(localEdge, connectData) {
    var id;
    id = localEdge.localId();
    this.edges.local[id] = {
      local: {
        mode: this.context.mode,
        globalId: this.context.globalId()
      },
      remote: connectData
    };
    if (this.context.mode !== 'root') {
      this.context.uplink.getPublisher()('event:edge:connect', this.edges.local[id]);
    }
    if (this.context.onChange && this.context.onChange.localEdge) {
      return this.context.onChange.localEdge(this.context, localEdge);
    }
  };

  Tree.prototype.insertRemote = function(connectData) {
    var id;
    id = connectData.local.globalId;
    this.edges.remote[id] = connectData;
    if (this.context.mode !== 'root') {
      this.context.uplink.getPublisher()('event:edge:connect', this.edges.remote[id]);
    }
    if (this.context.onChange && this.context.onChange.remoteEdge) {
      return this.context.onChange.remoteEdge(this.context, connectData);
    }
  };

  Tree.prototype.removeLocal = function(localEdge) {
    var id;
    id = localEdge.localId();
    if (!this.edges.local[id]) {
      return;
    }
    this.edges.local[id].disconnected = new Date();
    if (this.context.mode !== 'root') {
      this.context.uplink.getPublisher()('event:edge:disconnect', this.edges.local[id]);
    }
    if (this.context.onChange && this.context.onChange.localEdge) {
      return this.context.onChange.localEdge(this.context, localEdge);
    }
  };

  Tree.prototype.removeRemote = function(disconnectData) {
    var id;
    id = disconnectData.local.globalId;
    this.edges.remote[id] = disconnectData;
    if (this.context.mode !== 'root') {
      this.context.uplink.getPublisher()('event:edge:disconnect', this.edges.remote[id]);
    }
    if (this.context.onChange && this.context.onChange.remoteEdge) {
      return this.context.onChange.remoteEdge(this.context, disconnectData);
    }
  };

  return Tree;

})();

module.exports = Tree;
