// viewport/lib/viewport_proxy.js
// Generated by CoffeeScript 1.4.0
var ViewportProxy,
  __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

module.exports = ViewportProxy = (function() {

  function ViewportProxy() {
    this.bind = __bind(this.bind, this);
    this.viewports = {};
    this.index = {
      id2group: {}
    };
  }

  ViewportProxy.prototype.bind = function(socket) {
    var _this = this;
    socket.on('disconnect', function() {
      return _this.handleDisconnect(socket);
    });
    socket.on('viewport:connect', function(data) {
      return _this.handleConnect(socket, data);
    });
    return socket.on('viewport:event', function(data) {
      return _this.handleEvent(socket, data);
    });
  };

  ViewportProxy.prototype.handleConnect = function(socket, data) {
    var group, id, _base;
    id = socket.id;
    group = data.group;
    this.index.id2group[id] = group;
    (_base = this.viewports)[group] || (_base[group] = {
      sockets: {}
    });
    this.viewports[group].sockets[id] = socket;
    console.log('CONNECTED', data);
    console.log('viewports', this.viewports);
    console.log('index', this.index);
    return {
      connectresponse: 'TODO'
    };
  };

  ViewportProxy.prototype.handleDisconnect = function(socket) {
    var group, id;
    id = socket.id;
    group = this.index.id2group[id];
    delete this.viewports[group].sockets[id];
    delete this.index.id2group[id];
    console.log('DISCONNECTED', group);
    console.log('viewports', this.viewports);
    return console.log('index', this.index);
  };

  ViewportProxy.prototype.handleEvent = function(socket, data) {
    var group, id, peerId, _results;
    id = socket.id;
    group = this.index.id2group[id];
    if (!group) {
      return;
    }
    _results = [];
    for (peerId in this.viewports[group].sockets) {
      _results.push(this.viewports[group].sockets[peerId].emit('viewport:event', data));
    }
    return _results;
  };

  return ViewportProxy;

})();
